src/mapper.rs

```diff
@@ -0,0 +1,224 @@
+///
+/// This can remap frames one at a time, or the complete raw stacktrace.
+#[derive(Debug)]
+pub struct ProguardMapper<'s> {
+    classes: HashMap<&'s str, ClassMapping<'s>>,
+}
+
+impl<'s> From<&'s str> for ProguardMapper<'s> {
+    fn from(s: &'s str) -> Self {
+        let mapping = ProguardMapping::new(s.as_ref());
+        Self::new(mapping)
+    }
+}
+
+impl<'s> ProguardMapper<'s> {
+    /// Create a new ProguardMapper.
+    pub fn new(mapping: ProguardMapping<'s>) -> Self {
+        let mut classes = HashMap::new();
+        let mut class = ClassMapping {
+            original: "",
+            obfuscated: "",
+            members: BTreeMap::new(),
+        };
+
+        for record in mapping.iter().filter_map(Result::ok) {
+            match record {
+                MappingRecord::Class {
+                    original,
+                    obfuscated,
+                } => {
+                    if !class.original.is_empty() {
+                        classes.insert(class.obfuscated, class);
+                    }
+                    class = ClassMapping {
+                        original,
+                        obfuscated,
+                        members: BTreeMap::new(),
+                    }
+                }
+                MappingRecord::Method {
+                    original,
+                    obfuscated,
+                    original_class,
+                    line_mapping,
+                    ..
+                } => {
+                    // in case the mapping has no line records, we use `0` here.
+                    let (startline, endline) =
+                        line_mapping.as_ref().map_or((0, 0), |line_mapping| {
+                            (line_mapping.startline, line_mapping.endline)
+                        });
+                    let (original_startline, original_endline) =
+                        line_mapping.map_or((0, None), |line_mapping| {
+                            match line_mapping.original_startline {
+                                Some(original_startline) => {
+                                    (original_startline, line_mapping.original_endline)
+                                }
+                                None => (line_mapping.startline, Some(line_mapping.endline)),
+                            }
+                        });
+                    let members = class.members.entry(obfuscated).or_insert_with(|| vec![]);
+                    members.push(MemberMapping {
+                        startline,
+                        endline,
+                        original_class,
+                        original,
+                        original_startline,
+                        original_endline,
+                    });
+                }
+                _ => {}
+            }
+        }
+        if !class.original.is_empty() {
+            classes.insert(class.obfuscated, class);
+        }
+
+        Self { classes }
+    }
+
+    /// Remaps an obfuscated Class.
```

Comment:
_nit_: It's clear from the example, but might be worth pointing out that this needs to be the FQN.